sound/soc/sof/ipc4-topology.c

```diff
@@ -407,6 +407,39 @@ static void sof_ipc4_widget_update_kcontrol_module_id(struct snd_sof_widget *swi
 	}
 }
 
+static int
+sof_ipc4_update_card_components_string(struct snd_sof_widget *swidget,
+				       struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_sof_widget *pipe_widget = swidget->spipe->pipe_widget;
+	struct sof_ipc4_pipeline *pipeline = pipe_widget->private;
+	struct snd_soc_component *scomp = spcm->scomp;
+	struct snd_soc_card *card = scomp->card;
+
+	/*
+	 * Update the card's components list with hdmi-pt (passthrough) if
+	 * ChainDMA is enabled for HDMI PCM stream
+	 */
+	if (!strstr(spcm->pcm.pcm_name, "HDMI") || !pipeline->use_chain_dma)
+		return 0;
+
+	if (card->components && strstr(card->components, "hdmi-pt"))
+		return 0;
+
+	if (card->components)
+		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
+						  "%s hdmi-pt",
```

Comment:
@ujfalusi in the case that card->components is not NULL, dont you have to free it first before allocating a new one?

Reply:
we never do that in sound/soc/intel/boards/ , but I agree.
It looks like we always use devm_kasprintf() for card->components, but if some code path will drop the devm_ then devm_kfree() will scream...
Let me add this to the patch.

Reply:
@plbossart, @ranj063, @kv2019i: https://github.com/thesofproject/linux/pull/4928